base/math.jl

```diff
@@ -146,36 +146,51 @@ sqrt(x::Float32) = box(Float32,sqrt_llvm(unbox(Float32,x)))
 sqrt(x::Real) = sqrt(float(x))
 @vectorize_1arg Number sqrt
 
-hypot(x::Real, y::Real) = hypot(promote(float(x), float(y))...)
-function hypot{T<:AbstractFloat}(x::T, y::T)
-    x = abs(x)
-    y = abs(y)
-    if x < y
-        x, y = y, x
+"""
+    hypot(x, y)
+
+Compute the hypotenuse ``\\sqrt{x^2+y^2}`` avoiding overflow and underflow.
+"""
+hypot(x::Number, y::Number) = hypot(promote(x, y)...)
+function hypot{T<:Number}(x::T, y::T)
+    ax = abs(x)
+    ay = abs(y)
+    if ax < ay
+        ax, ay = ay, ax
     end
-    if x == 0
-        r = y/one(x)
+    if ax == 0
```

Comment:
For comparisons, you don't need to call `zero`. It's only useful when you want a specific type.